src/__experimental__/components/textarea/textarea.component.js

```diff
@@ -0,0 +1,172 @@
+    cols: PropTypes.number,
+    /** Adds disabled property */
+    disabled: PropTypes.bool,
+    /** Stop the user typing over the characterLimit */
+    enforceCharacterLimit: PropTypes.bool,
+    /** Allows the Textareas Height to change based on user input */
+    expandable: PropTypes.bool,
+    label: PropTypes.string,
+    onChange: PropTypes.func,
+    placeholder: PropTypes.string,
+    /** Adds readOnly property */
+    readOnly: PropTypes.bool,
+    /** The number of visible text lines for the control */
+    rows: PropTypes.number,
+    /** One of type of size to apply to the textarea */
+    size: PropTypes.oneOf(['small', 'medium', 'large']),
+    /** The value of the Textarea */
+    value: PropTypes.string,
+    /** Whether to display the character count message in red */
+    warnOverLimit: PropTypes.bool
+  }
+
+  static defaultProps = {
+    disabled: false,
+    expandable: false,
+    enforceCharacterLimit: true,
+    readOnly: false,
+    warnOverLimit: false
+  }
+
+  // Minimum height of the textarea
+  minHeight = 0;
+
+  /**
+   * A lifecycle method that is called after initial render.
+   * Allows access to refs and DOM to set expandable variables
+   */
+  componentDidMount() {
+    if (this.props.expandable) {
+      window.addEventListener('resize', this.expandTextarea);
+      // Set the min height to the initially rendered height.
+      // Without minHeight expandable textareas will only have
+      // one line when no content is present.
+      this.minHeight = this._input.clientHeight;
+
+      this.expandTextarea();
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.props.expandable) {
+      window.removeEventListener('resize', this.expandTextarea);
+    }
+  }
+
+  componentDidUpdate() {
+    const { expandable } = this.props;
+
+    if (expandable) {
+      this.expandTextarea();
+    }
+  }
+
+  expandTextarea = () => {
+    const textarea = this._input;
+
+    if (textarea.scrollHeight > this.minHeight) {
+      // Reset height to zero - IE specific
+      textarea.style.height = '0px';
+      // Set the height so all content is shown
+      textarea.style.height = `${Math.max(textarea.scrollHeight, this.minHeight)}px`;
+    }
+  }
+
+  get overLimit() {
+    const value = this.props.value || '';
+    return value.length > parseInt(this.props.characterLimit, 10);
+  }
+
+  get inputProps() {
+    const { ...props } = validProps(this);
+    props.rows = this.props.rows;
+    props.cols = this.props.cols;
+
+    if (this.props.characterLimit && this.props.enforceCharacterLimit) {
+      props.maxLength = this.props.characterLimit;
+    }
+
+    return props;
+  }
+
+  get characterCount() {
+    const value = this.props.value || '';
+    const { characterLimit, warnOverLimit } = this.props;
+
+    if (!characterLimit) { return null; }
+
+    return (
+      <CharacterCount
+        isOverLimit={ this.overLimit && warnOverLimit }
```

Comment:
I think that we should also consider `enforceCharacterLimit` - and pass `warnOverLimit` to characterCount only when `enforceCharacterLimit` is set to false

Reply:
![Zrzut ekranu 2019-04-16 o 13 58 42](https://user-images.githubusercontent.com/12142840/56207720-c4167600-604f-11e9-8f2b-6ff13eb2e32f.png)
I'm not sure if I want to change business logic here. If I change it I need add additional logic to slice text when characterLimit prop is changed.